mipengine/algorithms/logistic_regression.py

```diff
@@ -7,223 +7,245 @@
+    return coeff
 
-@udf
-def tensor_expit(t: TensorT[DT, ND]) -> TensorT[DT, ND]:
-    from scipy import special
 
-    result = special.expit(t)
-    return result
+# ~~~~~~~~~~~~~~~~~~~~~~~~ UDFs ~~~~~~~~~~~~~~~~~~~~~~~~~~ #
 
 
-# SQL UDF
-def tensor_mult(t1, t2):
-    result = t1 * t2
-    return result
+T = TypeVar("T")
+N = TypeVar("N")
+S = TypeVar("S")
 
 
-# SQL UDF
-def tensor_add(t1, t2):
-    result = t1 + t2
+@udf(n=literal(), return_type=tensor(float, 1))
+def zeros1(n):
+    result = numpy.zeros((n,))
     return result
 
 
-# SQL UDF
-def tensor_sub(t1, t2):
-    result = t1 - t2
-    return result
-
+@udf(rel=relation(S), return_type=tensor(float, 2))
+def relation_to_matrix(rel):
+    return rel
 
-# SQL UDF
-def tensor_div(t1, t2):
-    result = t1 / t2
-    return result
 
+@udf(rel=relation(S), return_type=tensor(float, 1))
+def relation_to_vector(rel):
+    return rel
 
-# SQL UDF
-def const_tensor_sub(const, t):
-    result = const - t
-    return result
 
+@udf(t=tensor(T, N), return_type=tensor(float, N))
+def tensor_expit(t):
+    from scipy import special
 
-# SQL UDF
-def mat_transp_dot_diag_dot_mat(M, d):
-    result = M.T @ numpy.diag(d) @ M
+    result = special.expit(t)
     return result
 
 
-# SQL UDF
-def mat_transp_dot_diag_dot_vec(M, d, v):
-    result = M.T @ numpy.diag(d) @ v
+@udf(vec=tensor(T, 1), return_type=tensor(T, 2))
+def diag(vec):
+    result = numpy.diag(vec)
     return result
 
 
-@udf
-def logistic_loss(v1: TensorT[DT, ND], v2: TensorT[DT, ND]) -> ScalarT(float):
+@udf(v1=tensor(T, N), v2=tensor(T, N), return_type=scalar(float))
+def logistic_loss(v1, v2):
     from scipy import special
 
     ll = numpy.sum(special.xlogy(v1, v2) + special.xlogy(1 - v1, 1 - v2))
     return ll
 
 
-@udf
-def tensor_max_abs_diff(t1: TensorT[DT, ND], t2: TensorT[DT, ND]) -> ScalarT(float):
+@udf(t1=tensor(T, N), t2=tensor(T, N), return_type=scalar(float))
+def tensor_max_abs_diff(t1, t2):
     result = numpy.max(numpy.abs(t1 - t2))
     return result
 
 
-@udf
-def mat_inverse(m: TensorT(Number, 2)) -> TensorT(float, 2):
+@udf(m=tensor(T, 2), return_type=tensor(float, 2))
+def mat_inverse(m):
     minv = numpy.linalg.inv(m)
     return minv
 
 
-# -------------------------------------------------------- #
-# Examples                                                 #
-# -------------------------------------------------------- #
-def true_run():
-    data = pd.read_csv("mipengine/algorithms/auxfiles/logistic_data.csv")
-    y = data["alzheimerbroadcategory"].to_numpy()
-    X = data[["lefthippocampus", "righthippocampus"]].to_numpy()
-    coeff = logistic_regression(y, X)
-    print(coeff)
+@udf(xs=merge_tensor(dtype=T, ndims=N), return_type=tensor(dtype=T, ndims=N))
+def sum_tensors(xs):
+    from functools import reduce
+
+    reduced = reduce(lambda a, b: a + b, xs)
+    reduced = numpy.array(reduced)
+    return reduced
 
 
-if __name__ == "__main__":
-    true_run()
+def test_logistic_regression():
```

Comment:
Did you forget this here?

Reply:
No, I use this to test drive the logistic regression algorithm. I will certainly need it in the future so I left it here. Sue me.